File: aquatrack_app/lib/core/models/water_profile.dart
import 'package:json_annotation/json_annotation.dart';

part 'water_profile.g.dart';

/// Enums cho water profile
enum Gender {
  @JsonValue('male')
  male,
  @JsonValue('female')
  female,
  @JsonValue('other')
  other
}

enum ActivityLevel {
  @JsonValue('sedentary')
  sedentary,
  @JsonValue('light')
  light,
  @JsonValue('moderate')
  moderate,
  @JsonValue('active')
  active,
  @JsonValue('very_active')
  veryActive
}

enum JobType {
  @JsonValue('office')
  office,
  @JsonValue('mixed')
  mixed,
  @JsonValue('outdoor')
  outdoor,
  @JsonValue('manual')
  manual
}

enum HealthCondition {
  @JsonValue('none')
  none,
  @JsonValue('diabetes')
  diabetes,
  @JsonValue('hypertension')
  hypertension,
  @JsonValue('neurological')
  neurological,
  @JsonValue('heart')
  heart,
  @JsonValue('pregnant')
  pregnant,
  @JsonValue('lactating')
  lactating,
  @JsonValue('gout')
  gout
}

enum VeggieIntake {
  @JsonValue('low')
  low,
  @JsonValue('medium')
  medium,
  @JsonValue('high')
  high
}

/// Water profile enum labels cho UI
@JsonSerializable()
class WaterProfileEnums {
  final Map<String, String> genders;
  final Map<String, String> activityLevels;
  final Map<String, String> jobTypes;
  final Map<String, String> healthConditions;
  final Map<String, String> veggieIntakes;

  WaterProfileEnums({
    required this.genders,
    required this.activityLevels,
    required this.jobTypes,
    required this.healthConditions,
    required this.veggieIntakes,
  });

  factory WaterProfileEnums.fromJson(Map<String, dynamic> json) =>
      _$WaterProfileEnumsFromJson(json);

  Map<String, dynamic> toJson() => _$WaterProfileEnumsToJson(this);
}

/// Water profile update request
@JsonSerializable()
class WaterProfileUpdate {
  final Gender? gender;
  final int? age;
  final int? height; // cm
  final double? weight; // kg
  final ActivityLevel? activityLevel;
  final JobType? jobType;
  final List<HealthCondition>? healthConditions;
  final VeggieIntake? veggieIntake;
  final int? coffeeCupsPerDay;
  final int? alcoholUnitsPerDay;

  WaterProfileUpdate({
    this.gender,
    this.age,
    this.height,
    this.weight,
    this.activityLevel,
    this.jobType,
    this.healthConditions,
    this.veggieIntake,
    this.coffeeCupsPerDay,
    this.alcoholUnitsPerDay,
  });

  factory WaterProfileUpdate.fromJson(Map<String, dynamic> json) =>
      _$WaterProfileUpdateFromJson(json);

  Map<String, dynamic> toJson() => _$WaterProfileUpdateToJson(this);
}

/// Water calculation breakdown
@JsonSerializable()
class WaterCalculationBreakdown {
  final int baseMl;
  final int activityAdd;
  final int jobAdd;
  final int healthAdd;
  final int veggieAdd;
  final int coffeeAdd;
  final int alcoholAdd;

  WaterCalculationBreakdown({
    required this.baseMl,
    required this.activityAdd,
    required this.jobAdd,
    required this.healthAdd,
    required this.veggieAdd,
    required this.coffeeAdd,
    required this.alcoholAdd,
  });

  factory WaterCalculationBreakdown.fromJson(Map<String, dynamic> json) =>
      _$WaterCalculationBreakdownFromJson(json);

  Map<String, dynamic> toJson() => _$WaterCalculationBreakdownToJson(this);
}

/// Water calculation response
@JsonSerializable()
class WaterCalculationResponse {
  final int totalMl;
  final double dailyGoalL;
  final int dailyGoalCups;
  final WaterCalculationBreakdown breakdown;
  final bool hasWarnings;
  final String? warningMessage;
  final DateTime calculatedAt;

  WaterCalculationResponse({
    required this.totalMl,
    required this.dailyGoalL,
    required this.dailyGoalCups,
    required this.breakdown,
    required this.hasWarnings,
    this.warningMessage,
    required this.calculatedAt,
  });

  factory WaterCalculationResponse.fromJson(Map<String, dynamic> json) =>
      _$WaterCalculationResponseFromJson(json);

  Map<String, dynamic> toJson() => _$WaterCalculationResponseToJson(this);
}

/// Water profile response
@JsonSerializable()
class WaterProfileResponse {
  final Gender? gender;
  final int? age;
  final int? height;
  final double? weight;
  final ActivityLevel? activityLevel;
  final JobType? jobType;
  final List<HealthCondition>? healthConditions;
  final VeggieIntake? veggieIntake;
  final int? coffeeCupsPerDay;
  final int? alcoholUnitsPerDay;
  final bool profileComplete;
  final int? calculatedDailyGoalMl;
  final DateTime? formulaLastUpdated;

  WaterProfileResponse({
    this.gender,
    this.age,
    this.height,
    this.weight,
    this.activityLevel,
    this.jobType,
    this.healthConditions,
    this.veggieIntake,
    this.coffeeCupsPerDay,
    this.alcoholUnitsPerDay,
    required this.profileComplete,
    this.calculatedDailyGoalMl,
    this.formulaLastUpdated,
  });

  factory WaterProfileResponse.fromJson(Map<String, dynamic> json) =>
      _$WaterProfileResponseFromJson(json);

  Map<String, dynamic> toJson() => _$WaterProfileResponseToJson(this);
}

/// User summary response cho B5 Review screen
@JsonSerializable()
class UserSummaryResponse {
  final String genderAge; // "Nam - 28 tuổi"
  final String heightWeight; // "168 cm - 60 kg"
  final String activity; // "Vừa phải"
  final String job; // "Văn phòng"

  UserSummaryResponse({
    required this.genderAge,
    required this.heightWeight,
    required this.activity,
    required this.job,
  });

  factory UserSummaryResponse.fromJson(Map<String, dynamic> json) =>
      _$UserSummaryResponseFromJson(json);

  Map<String, dynamic> toJson() => _$UserSummaryResponseToJson(this);
}

/// Extension methods cho enum labels
extension GenderExtension on Gender {
  String get label {
    switch (this) {
      case Gender.male:
        return 'Nam';
      case Gender.female:
        return 'Nữ';
      case Gender.other:
        return 'Khác';
    }
  }
}

extension ActivityLevelExtension on ActivityLevel {
  String get label {
    switch (this) {
      case ActivityLevel.sedentary:
        return 'Ít vận động';
      case ActivityLevel.light:
        return 'Nhẹ nhàng';
      case ActivityLevel.moderate:
        return 'Vừa phải';
      case ActivityLevel.active:
        return 'Năng động';
      case ActivityLevel.veryActive:
        return 'Rất năng động';
    }
  }
}

extension JobTypeExtension on JobType {
  String get label {
    switch (this) {
      case JobType.office:
        return 'Văn phòng';
      case JobType.mixed:
        return 'Hỗn hợp';
      case JobType.outdoor:
        return 'Ngoài trời';
      case JobType.manual:
        return 'Tay chân';
    }
  }
}

extension HealthConditionExtension on HealthCondition {
  String get label {
    switch (this) {
      case HealthCondition.none:
        return 'Không có';
      case HealthCondition.diabetes:
        return 'Tiểu đường';
      case HealthCondition.hypertension:
        return 'Cao huyết áp';
      case HealthCondition.neurological:
        return 'Bệnh thần kinh';
      case HealthCondition.heart:
        return 'Tim mạch';
      case HealthCondition.pregnant:
        return 'Đang mang thai';
      case HealthCondition.lactating:
        return 'Đang cho con bú';
      case HealthCondition.gout:
        return 'Gout';
    }
  }
}

extension VeggieIntakeExtension on VeggieIntake {
  String get label {
    switch (this) {
      case VeggieIntake.low:
        return 'Ít (< 1 phần/ngày)';
      case VeggieIntake.medium:
        return 'Vừa (1-2 phần/ngày)';
      case VeggieIntake.high:
        return 'Nhiều (3+ phần/ngày)';
    }
  }
}

File: aquatrack_backend/app/schemas/water_profile.py
"""
Pydantic schemas for water profile and calculation
"""

from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, validator

from app.services.water_formula_service import (ActivityLevel, Gender,
                                                HealthCondition, JobType,
                                                VeggieIntake)


class WaterProfileBase(BaseModel):
    """Base schema for water profile"""

    # B1 - Body
    gender: Optional[Gender] = None
    age: Optional[int] = None
    height: Optional[int] = None  # cm
    weight: Optional[float] = None  # kg

    # B2 - Lifestyle
    activity_level: Optional[ActivityLevel] = None
    job_type: Optional[JobType] = None

    # B3 - Health
    health_conditions: List[HealthCondition] = ["none"]

    # B4 - Diet
    veggie_intake: Optional[VeggieIntake] = None
    coffee_cups_per_day: int = 0
    alcohol_units_per_day: int = 0

    @validator("health_conditions")
    def validate_health_conditions(cls, v):
        """Ensure health_conditions is always a list"""
        if not v:
            return ["none"]
        if "none" in v and len(v) > 1:
            # If "none" is selected with others, remove "none"
            return [cond for cond in v if cond != "none"]
        return v

    @validator("age")
    def validate_age(cls, v):
        if v is not None and not (1 <= v <= 120):
            raise ValueError("Tuổi phải từ 1 đến 120")
        return v

    @validator("height")
    def validate_height(cls, v):
        if v is not None and not (130 <= v <= 210):
            raise ValueError("Chiều cao phải từ 130cm đến 210cm")
        return v

    @validator("weight")
    def validate_weight(cls, v):
        if v is not None and not (30 <= v <= 150):
            raise ValueError("Cân nặng phải từ 30kg đến 150kg")
        return v

    @validator("coffee_cups_per_day")
    def validate_coffee(cls, v):
        if not (0 <= v <= 10):
            raise ValueError("Số cốc cà phê phải từ 0 đến 10")
        return v

    @validator("alcohol_units_per_day")
    def validate_alcohol(cls, v):
        if not (0 <= v <= 10):
            raise ValueError("Số đơn vị rượu bia phải từ 0 đến 10")
        return v


class WaterProfileCreate(WaterProfileBase):
    """Schema for creating water profile"""


class WaterProfileUpdate(WaterProfileBase):
    """Schema for updating water profile"""


class WaterCalculationBreakdown(BaseModel):
    """Breakdown of water calculation components"""

    base_ml: int
    activity_add: int
    job_add: int
    health_add: int
    veggie_add: int
    coffee_add: int
    alcohol_add: int


class WaterCalculationResponse(BaseModel):
    """Response schema for water calculation"""

    total_ml: int
    daily_goal_l: float
    daily_goal_cups: int
    breakdown: WaterCalculationBreakdown
    has_warnings: bool
    warning_message: Optional[str] = None
    calculated_at: datetime


class WaterProfileResponse(WaterProfileBase):
    """Response schema for water profile"""

    profile_complete: bool
    calculated_daily_goal_ml: Optional[int] = None
    formula_last_updated: Optional[datetime] = None

    class Config:
        from_attributes = True


class UserSummaryResponse(BaseModel):
    """User summary for B5 Review screen"""

    gender_age: str  # "Nam - 28 tuổi"
    height_weight: str  # "168 cm - 60 kg"
    activity: str  # "Vừa phải"
    job: str  # "Văn phòng"

    class Config:
        from_attributes = True


# Enums for frontend display
class WaterProfileEnums(BaseModel):
    """Available enum values for frontend dropdowns"""

    genders: Dict[str, str] = {"male": "Nam", "female": "Nữ", "other": "Khác"}

    activity_levels: Dict[str, str] = {
        "sedentary": "Ít vận động",
        "light": "Nhẹ nhàng",
        "moderate": "Vừa phải",
        "active": "Năng động",
        "very_active": "Rất năng động",
    }

    job_types: Dict[str, str] = {
        "office": "Văn phòng",
        "mixed": "Hỗn hợp",
        "outdoor": "Ngoài trời",
        "manual": "Tay chân",
    }

    health_conditions: Dict[str, str] = {
        "none": "Không có",
        "diabetes": "Tiểu đường",
        "hypertension": "Cao huyết áp",
        "neurological": "Bệnh thần kinh",
        "heart": "Tim mạch",
        "pregnant": "Đang mang thai",
        "lactating": "Đang cho con bú",
        "gout": "Gout",
    }

    veggie_intakes: Dict[str, str] = {
        "low": "Ít (< 1 phần/ngày)",
        "medium": "Vừa (1-2 phần/ngày)",
        "high": "Nhiều (3+ phần/ngày)",
    }
